Moving a saved notebook makes it lose track of the kernel

dhruvisompura opened the pull request:
Addresses #9089 (and #9106 in the sense that we have new UX for surfacing the cwd to users when there is a mismatch).
  
When a notebook file is moved or **an untitled notebook is saved**, the interpreter session for the notebook continues to use the old working directory. This PR attempts to address this problem by detecting when there is a directory mismatch and then surfacing a button in the action bar of a notebook for the user to resolve this issue. 

**NOTE**: We no longer show the update working directory dialog when an untitled notebook is saved. I was chatting with @austin3dickey about the issue and it made more sense for us to consolidate all paths. 

We are not automatically updating the working directory because we think there are scenarios where a user wouldn't want to change the working directory. We may want to revisit the experience for an untitled notebook (maybe we should automatically update the working directory for this scenario?).

This PR also addresses the problem where moving a positron notebook to a different directory would cause the notebook editor to close! This problem needed to be fixed before the new UX for #9089 could be put forth. The cause of this issue was a missing implementation for `rename` on `PositronNotebookEditorInput`. This method is called when a notebook file is moved to a different location. The implementation return value determines how the editor resource is handled. No implementation means there is no editor for the moved file.

--

The command ID to update the working directory of a notebook is: `positronNotebook.updateWorkingDirectory`. We want to allow this command to run if the active editor is a notebook (Positron or built-in).

For built-in notebooks, this action will be accessible via the command palette but we don't want to show it in the action bar when there is a working directory mismatch. For positron notebooks, we only want to show the action in the action bar when the working directory does not match the notebook location.

When a notebook's working directory differs from its file location, an alert icon (⚠️) appears in the editor action bar. Clicking the icon will present the user with a quick pick to confirm updating the working directory.
    
The code that checks for a directory mismatch _SHOULD_ fire for Positron Notebooks:
- When a notebook is opened and a runtime session is attached
- When a runtime session starts or becomes ready
- When the working directory changes (including via code execution as described below)
  - Python: `os.chdir('/some/path')` or `%cd /some/path`
  - R: `setwd('/some/path')`
- When the notebook file is moved
- When untitled notebook is saved

--

**Save Untitled Notebook**

https://github.com/user-attachments/assets/5855ef9e-c664-49b2-82fb-9560d93c3bc1

**Move Existing Notebook**

https://github.com/user-attachments/assets/dd7570d9-8a57-48a6-9c24-06d71763b6bb

**No-op when working directory has not changed**

https://github.com/user-attachments/assets/711cb1a4-8e3e-4041-9199-bb924fd0134d

**Switching Editor Tabs Reflects Correct Working Directory Mismatch State**

https://github.com/user-attachments/assets/5f32d55c-ee37-4047-8ee9-9495f9716ba3

**Changing Working Directory With Code**

https://github.com/user-attachments/assets/f43181fc-f59b-4b50-a322-2a81d9ecccd9

https://github.com/user-attachments/assets/cdb9991e-3dc8-47cb-9a9b-5b870cda719a

**Update Working Directory For Built-In Notebooks**

https://github.com/user-attachments/assets/69f2c717-fa5f-44ef-911d-4d83ac70c2f1

**Keep Kernel Attached For Built-In Notebooks**

BEFORE - new runtime session gets created

https://github.com/user-attachments/assets/cb3e389a-6678-42af-8a0c-4d5c519bcae3

AFTER - we reuse the same runtime session

https://github.com/user-attachments/assets/356e1705-058a-4d71-8948-24d59b10df23



### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

The command ID to update the working directory of a notebook is: `positronNotebook.updateWorkingDirectory`.

There are different ways to change the working directory as described above in the PR and we should test against those scenarios! There's lots of details in the PR description above that should help for testing.

Note that the following setting should be _REMOVED_ when testing if the kernel in a VSC notebook so that when the notebook is moved to a different directory, it opens with the VSC notebook editor and not the Positron Notebook Editor (I think this is expected/makes sense since you've set a default editor for that file type):
```
"workbench.editorAssociations": {
        "*.ipynb": "workbench.editor.positronNotebook"
    },
```

@:positron-notebooks

<!--
  Positron team members: please add relevant e2e test tags, so the tests can be
  run when you open this pull request.

  - Instructions: https://github.com/posit-dev/positron/blob/main/test/e2e/README.md#pull-requests-and-test-tags
  - Available tags: https://github.com/posit-dev/positron/blob/main/test/e2e/infra/test-runner/test-tags.ts
-->


<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

dhruvisompura committed:
WIP

dhruvisompura committed:
Create a quick pick menu with a description

dhruvisompura committed:
use helper to get notebook instance

dhruvisompura committed:
Update text for quickpick

dhruvisompura committed:
Move common directory functions to a util file

Use in new action

dhruvisompura committed:
Refactor runtime session service to use utils

dhruvisompura committed:
Notify user if action no ops

dhruvisompura committed:
Implement rename so notebook stays open on move

dhruvisompura committed:
Remove dialog prompt for untitled notebook files

dhruvisompura committed:
Fix action so it fires

dhruvisompura committed:
Move notebook directory utils file

dhruvisompura committed:
Add context key and update it

dhruvisompura committed:
Fix issue where action wouldn't show up in action bar

Needed to update directory mismatch key when switching tabs and give editor action bar access to scoped context keys for notebook

dhruvisompura committed:
Fix leaky disposable warning from quickpick

dhruvisompura committed:
Notify user on success

dhruvisompura committed:
Fix error surfaced in dev tools console

dhruvisompura committed:
Clear context key after updating working directory

dhruvisompura committed:
Use absolute path when updating working directory

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookEditor.tsx:
```diff
@@ -328,6 +326,10 @@ export class PositronNotebookEditor extends AbstractEditorWithViewState<INoteboo
+	override get scopedContextKeyService(): IContextKeyService | undefined {
+		return this.notebookInstance?.scopedContextKeyService ?? this._containerScopedContextKeyService;
+	}
+
```
I was running into the issue of the editor action bar not having access to the scoped context keys for the notebook editor. Apparently, this is a way for providing access to the scoped context keys. This was a suggestion from claude so please scrutinize away!

nstrayer replied:
Oh that's good! I had been annoyed by this before.

nstrayer reviewed (dismissed):
This turned out really nice!

I wonder if the warning should be more in-your-face. Also potentially with the ability for the user to dismiss it if they know what they're doing. Ultimately that's design work though so probably best left to another PR.

austin3dickey commented on src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookEditorInput.ts:
```diff
@@ -381,4 +366,42 @@ export class PositronNotebookEditorInput extends EditorInput {
+	 * @param target The new URI for the notebook
+	 * @returns IMoveResult to keep the editor open with the new resource
+	 */
+	override async rename(group: GroupIdentifier, target: URI): Promise<IMoveResult | undefined> {
```
I tried out VS Code notebooks too, and while I expected there not to be an icon, it seems that we're still losing state when moving them (i.e. https://github.com/posit-dev/positron/issues/9089). Are we interested in 
- calling `updateNotebookSessionUri()` during `rename()`
- letting the new Update Working Directory command work

for those notebooks too?

dhruvisompura replied:
> Oh yeah I don't think it's worth trying to include the icon 😅 For the second bullet I was just wondering if it's easy to make the Command Palette command work for VSC notebooks without the icon. If not that's fine.

We could definitely make the command from the command palette work for VSC notebooks! We just need to change the `when` clause to be either an active positron notebook or VSC notebook. I think there is an existing context key that we can use.

austin3dickey reviewed (commented):
Awesome! Works great. I tried out Windows and Web too, which have given me problems in the past.

>We are not automatically updating the working directory because we think there are scenarios where a user wouldn't want to change the working directory. We may want to revisit the experience for an untitled notebook (maybe we should automatically update the working directory for this scenario?).

Yeah, it always felt like overreaching to automatically do this without confirmation from the user. I feel like the user will want to change the WD in almost all use cases, but it seems nice to basically ask permission first.

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/UpdateNotebookWorkingDirectoryAction.ts:
```diff
@@ -0,0 +1,232 @@
+				controlType: 'button',
+				displayTitle: false
+			},
+			icon: ThemeIcon.fromId('alert'),
```
We don't have a filled version of this icon unfortunately and changing the outline color to yellow would probably have contrast issues with the existing icon. We could maybe try orange (red feels too aggressive) but then I'm worried it would be too "attention grabby" for folks who don't actually want to take this action.

We could consider this as polish work for the future but probably not priority for this PR. Ideally, we also add a way for users to dismiss this action from the editor bar.

austin3dickey replied:
>I'm worried it would be too "attention grabby" for folks who don't actually want to take this action

I totally get this, but also if they ever restart the notebook kernel (e.g. when restarting Positron) the WD is going to get updated automatically. So it could be worth being more attention grabby. But happy to defer this as design polish!

I'm thinking this mismatch is on a similar level to the "These extensions want to relaunch the Terminal to contribute to its environment" notification:

<img width="259" height="33" alt="image" src="https://github.com/user-attachments/assets/b68edabf-3461-4f33-8d0e-110b62697e17" />

seeM reviewed (dismissed):
This is great!

dhruvisompura committed:
Remove unnecessary idle check

dhruvisompura committed:
Move context key logic into active context key manager

dhruvisompura committed:
Remove context key logic from notebook instance

dhruvisompura committed:
Use context key from active notebook context key manager

dhruvisompura committed:
Merge branch 'main' into positron-nb-cwd-work-2

 Conflicts:
	src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookEditorInput.ts
	src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts
	src/vs/workbench/contrib/positronNotebook/browser/positronNotebook.contribution.ts
	src/vs/workbench/services/runtimeSession/common/runtimeSession.ts

dhruvisompura committed:
Fix error from merge: get runtime session object for notebook

We removed the runtime session object on the notebook instance and need to change
how we fetch the session

dhruvisompura committed:
Add kernel selection logic to rename

dhruvisompura committed:
Fix random ESLint error

dhruvisompura committed:
Support VS Code notebooks with action

dhruvisompura committed:
Keep track of kernel when moving built-in notebooks

dhruvisompura committed:
Update unit test

seeM reviewed (approved):
CI seems to be acting up so I reran failed, but otherwise this looks great ty!
